pkg/deployments/readiness.go

```diff
@@ -0,0 +1,65 @@
+package deployments
+
+import (
+	"context"
+	"errors"
+	"log"
+	"net/http"
+	"sync/atomic"
+	"time"
+
+	appsv1 "k8s.io/api/apps/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type k8sClient interface {
+	List(ctx context.Context, list client.ObjectList, opts ...client.ListOption) error
+}
+type Readiness struct {
+	ready       atomic.Bool
+	k8sClient   k8sClient
+	namespace   string
+	deployments *Manager
+	interval    time.Duration // await deployments loaded on startup
+}
+
+func NewReadiness(k8sClient k8sClient, namespace string, deployments *Manager, interval time.Duration) *Readiness {
+	return &Readiness{k8sClient: k8sClient, namespace: namespace, deployments: deployments, interval: interval}
+}
+
+func (h *Readiness) StateLoaded(r *http.Request) error {
+	if h.ready.Load() {
+		return nil
+	}
+	return errors.New("not ready")
+}
+
+func (h *Readiness) WatchModelBackendsLoaded() {
```

Comment:
This is called by main once and loops until all model backend deployments are in the deployment manager loaded. This may not work well when there are a lot of concurrent create/ deletes of deployments